### ISSUE TITLE
Disaster recovery

### ISSUE THREAD

**Issue by djjudas21** [democratic-csi/democratic-csi#273]
Another "support" type question for @travisghansen :slightly_smiling_face: 

Suppose I have a Kubernetes cluster with Democratic CSI and a TrueNAS appliance that has got a number of volumes provisioned. Suppose my cluster is a total loss for some reason and I have to rebuild all my nodes from scratch and I lose my etcd database which knows about all the PVCs and PVs.

Is there a way to start with a brand new cluster, install Democratic CSI again and somehow import the existing volumes, and recreate the PV objects in Kubernetes?

Thanks

**Comment by travisghansen** [comment]
Yikes :( sorry to hear about your supposition.

The short answer is no, nothing has been built.

The long answer is, it's hypothetically possible but would be a fair amount of work. 2 resources for things of similar nature are:

- https://github.com/democratic-csi/democratic-csi/blob/master/bin/k8s-csi-cleaner
- https://github.com/democratic-csi/democratic-csi/blob/master/contrib/freenas-provisioner-to-democratic-csi.sh

In short, the required data to reconstruct things is available as zfs props on the zvols/dataset. The csi `ListVolumes` method (see link #1) will return most of that data for each volume. With some effort you could iterate the `ListVolumes` call and then reconstruct the PVs. In general the driver knows little (or nothing) about PVCs however so the linkage would/could be challenging but not impossible.

**Comment by djjudas21** [comment]
I'm hoping it won't come to total loss of the cluster, but I'm currently talking to some of the MicroK8s developers about a bug which appears to have corrupted my cluster database (they're using dqlite rather than etcd, but same thing) and I can only get it to run on one node, read-only. Hoping for a method of restoring it and recovering the cluster, but I'm also preparing for a full rebuild.

I think I only have about 20 volumes on TrueNAS so it may be possible to recover them manually using your hints. I'm sure in the past I've manually re-associated PVs with PVCs, so it's mostly a case of recreating the PV objects. I'm a little bit hazy on the details of how PVs and CSI work, but what if I could manually recreate the PV objects in Kubernetes from manifests and reinstall Democratic CSI? Would it have enough info there to know that the PV relates to a volume?

It also might be easier to just let the new cluster provision new volumes, then `ssh` onto the NAS and `mv` content from the old volume to the new one, but that seems like a hack.

**Comment by travisghansen** [comment]
Yeah creating manifests is exactly what you will need to do (either applying yaml or using a k8s client sdk to do so). The format is not terrible and as I say, most of the information is available from the `ListVolumes` method. Of course you could use the freenas api (or just `zfs` commands) to get most of the data as well.

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: org.democratic-csi.nfs
  finalizers:
  - kubernetes.io/pv-protection
  name: pvc-195bdd1f-bf4c-11e8-b1e1-e0d55e88928e
spec:
  accessModes:
  - ReadWriteMany
  capacity:
    storage: 10Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: lachoneus-wordpress-wordpress
    namespace: lachoneus
  csi:
    controllerExpandSecretRef:
      name: controller-expand-secret-zfs-nfs-democratic-csi-zfs-nfs
      namespace: democratic-csi
    controllerPublishSecretRef:
      name: controller-publish-secret-zfs-nfs-democratic-csi-zfs-nfs
      namespace: democratic-csi
    driver: org.democratic-csi.nfs
    fsType: nfs
    nodePublishSecretRef:
      name: node-publish-secret-zfs-nfs-democratic-csi-zfs-nfs
      namespace: democratic-csi
    nodeStageSecretRef:
      name: node-stage-secret-zfs-nfs-democratic-csi-zfs-nfs
      namespace: democratic-csi
    # these are available via the ListVolumes method
    volumeAttributes:
      node_attach_driver: nfs
      provisioner_driver: freenas-nfs
      server: bitness.lan
      share: /mnt/pool0/k8s/on/v/pvc-195bdd1f-bf4c-11e8-b1e1-e0d55e88928e
    volumeHandle: pvc-195bdd1f-bf4c-11e8-b1e1-e0d55e88928e
  persistentVolumeReclaimPolicy: Delete
  storageClassName: zfs-nfs
  volumeMode: Filesystem
```

**Comment by djjudas21** [comment]
I can confirm that it was sufficient to export the PVs to YAML (e.g. `kubectl get pv -o yaml > pv.yaml` and reimport them on the "new" cluster (e.g. `kubectl apply -f pv.yaml`), provided I removed the `.spec.claimRef.uid` field (and left the other `claimRef` values). The PVs then show up in the new cluster as `Available`.

This allowed me to simply redeploy my applications and when new PVCs were created, they immediately matched the PVs and were bound.

I didn't need to manually retrieve anything from ZFS etc.

Hope this helps anyone else who has been an idiot and broken their cluster :wink:

**Comment by TheAceMan** [comment]
I have been testing some cluster deployment automation, this will allow me to retain the storage after a rebuild. I can export the PV and clear all uid values and apply after the rebuild. Is there any potential issue with this approach or a better way to do this with a cluster rebuild? I think I could also use snapshots but this seems more straight forward and does not require a snapshot.

I have tested within my existing cluster by deleting the PVC/PV but have not torn it down for a clean rebuild test yet.
